arch/arm/src/xmc4/xmc4_pwm.c

```diff
@@ -0,0 +1,626 @@
+/****************************************************************************
+ * arch/arm/src/xmc4/xmc4_pwm.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <nuttx/arch.h>
+#include <arch/board/board.h>
+
+#include "arm_internal.h"
+#include "chip.h"
+#include "xmc4_pwm.h"
+#include "xmc4_ccu4.h"
+#include "hardware/xmc4_ccu4.h"
+#include "xmc4_gpio.h"
+#include "hardware/xmc4_scu.h"
+#include "hardware/xmc4_pinmux.h"
+#include "xmc4_clockconfig.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* This structure represents the state of one PWM timer */
+
+struct xmc4_pwm_s
+{
+    const struct pwm_ops_s *ops; /* PWM operations */
+    uint8_t module;              /* CCU4x Module number {0,...,4} */
+    uint8_t slice;               /* CC4y Slice number {0,...,4} */
+    uint8_t prescaler;           /* Clock division for t_DTS */
+    uint32_t frequency;          /* Current frequency setting */
+    uint32_t duty;               /* Current frequency setting */
+    uint32_t base;               /* The base address of the Slice */
+    uint32_t f_tclk;             /* The frequency of the module clock that drives the timer module*/
+    uint32_t outgpio;
+};
```

Comment:
Please fix the spacing, should be two spaces in the new line below and after "{"